Interactive docker image

indy-3rdman opened the pull request:
This PR contains a build script, Dockerfile(s), README.md and supporting files to create a docker image that can run .NET for Apache Spark in an interactive Jupyter notebook. 

An initial description for the interactive image, along with the folder structure can be found here: https://github.com/indy-3rdman/spark/tree/interactive-docker-image/docker/images/interactive

MichaelSimons commented on docker/images/interactive/apache-spark/Dockerfile:
```diff
@@ -0,0 +1,30 @@
+    && rm spark-${SPARK_VERSION}-bin-hadoop${HADOOP_VERSION}.tgz \
+    && chmod 755 /usr/local/bin/start-spark-debug.sh \
+    && chown -R ${NB_UID} ${HOME} \
+    && cd ${HOME}/nb
```
This doesn't seem necessary.

MichaelSimons reviewed (commented):
I mainly looked at the dotnet-interactive/Dockerfile.  I left a few suggestions/questions that may help simplify the Dockerfile.

indy-3rdman commented:
@MichaelSimons, thank you very much for reviewing the Dockerfile and your valuable comments. I've just updated the PR to reflect the changes required for dotnet.spark version 1.0.0. This should also include an updated version of the Dockerfile, addressing your comments.

MichaelSimons commented on docker/images/interactive/build.sh:
```diff
@@ -8,12 +8,17 @@ set -o nounset   # abort on unbound variable
 readonly image_repository='3rdman'
-readonly supported_apache_spark_versions=("2.3.3" "2.3.4" "2.4.0" "2.4.1" "2.4.3" "2.4.4" "2.4.5" "2.4.6")
-readonly supported_dotnet_spark_versions=("0.9.0" "0.10.0" "0.11.0" "0.12.1")
+readonly supported_apache_spark_versions=(
```
This may be a question for Spark team.  Thoughts on how to keep this version list up-to-date and other versions included in this script up-to-date?  It feels like there should be long term plans for getting this updated "automatically" as part of the release process.  Without this they will become stale and/or be a maintenance burden.

indy-3rdman replied:
Agreed

MichaelSimons reviewed (commented):
@indy-3rdman, Thanks for addressing my comments. I took another look and had a few more suggestions and questions.

indy-3rdman commented on docker/images/interactive/dotnet-interactive/Dockerfile:
```diff
@@ -21,24 +14,30 @@ USER root
     && apt-get install -y --no-install-recommends \
-        dialog apt-utils wget ca-certificates openjdk-8-jdk bash software-properties-common supervisor unzip socat net-tools vim \
-        libc6 libgcc1 libgssapi-krb5-2 libicu60 libssl1.1 libstdc++6 zlib1g \
+        apt-utils \
```
This should be cleaned up now. Java obviously is required by spark.

indy-3rdman commented on docker/images/interactive/dotnet-interactive/Dockerfile:
```diff
@@ -21,24 +14,30 @@ USER root
+        libgssapi-krb5-2 \
+        libicu60 \
+        libssl1.1 \
+        libstdc++6 zlib1g \
```
Should be in a separate line as well now.

MichaelSimons commented on docker/images/interactive/dotnet-interactive/Dockerfile:
```diff
@@ -2,31 +2,24 @@ FROM jupyter/base-notebook:ubuntu-18.04
-ENV PATH="${PATH}:${HOME}/.dotnet/tools"
-
-ENV DOTNET_RUNNING_IN_CONTAINER=true \
+ENV DOTNET_CORE_VERSION=$DOTNET_CORE_VERSION \
```
Per the [Dockerfile Best Practices](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#sort-multi-line-arguments),  sort multi-line instructions to improve readability where possible (e.g. cross dependencies)

MichaelSimons commented on docker/images/interactive/dotnet-spark/Dockerfile:
```diff
@@ -4,28 +4,31 @@ FROM dotnet-spark-base-interactive:$DOTNET_SPARK_VERSION
+    && cd / \
+    && echo "\nDownloading spark-${SPARK_VERSION}-bin-hadoop${HADOOP_VERSION}.tgz ..." \
+    && wget -q https://archive.apache.org/dist/spark/spark-${SPARK_VERSION}/spark-${SPARK_VERSION}-bin-hadoop${HADOOP_VERSION}.tgz \
     && tar -xvzf spark-${SPARK_VERSION}-bin-hadoop${HADOOP_VERSION}.tgz \
```
You can extract to the spark directory with a single instruction which would eliminate the need for `mv`

indy-3rdman replied:
I assume you mean to use tar with --directory. But wouldn't that required that the directory exist already? In that case I'd have to add a mkdir first.

MichaelSimons replied:
Your correct, I missed what was happening here.  Please ignore my comment.

MichaelSimons commented on docker/images/interactive/dotnet-spark/Dockerfile:
```diff
@@ -4,28 +4,31 @@ FROM dotnet-spark-base-interactive:$DOTNET_SPARK_VERSION
+RUN cd /dotnet/HelloSpark \
+    && dotnet build \
+    && cp /dotnet/HelloSpark/bin/Debug/netcoreapp${DOTNET_CORE_VERSION}/microsoft-spark-*.jar ${HOME}/ \
+    && rm -rf /dotnet/HelloSpark \
```
The unfortunate consequence of this pattern is that HelloSpark remains in the image as a result of obtaining it via COPY.  This is not desirable.  Is there a way this can be generated during the Docker build or can it be a published tarball so that is can get copied and deleted within a single Dockerfile instruction?

indy-3rdman replied:
Thanks again @MichaelSimons for your great feedback! Just creating a dummy project during the build process now.

indy-3rdman commented on docker/images/interactive/dotnet-interactive/Dockerfile:
```diff
@@ -21,24 +14,30 @@ USER root
+        openjdk-8-jdk \
+        software-properties-common \
+        unzip \
     && wget -q --show-progress --progress=bar:force:noscroll https://packages.microsoft.com/config/ubuntu/18.04/packages-microsoft-prod.deb -O packages-microsoft-prod.deb \
```
That raises an interesting point about the purpose of the Dockerfile(s). As far as I am aware, the focus at the moment is to enable an user to build the image(s) her/himself, instead of automating the image build process. That's why I thought it would be useful to show the download progress. Now, for small downloads that doesn't really matter and I therefore have removed it. However, I have added the following line to the dotnet-spark/Dockerfile

`&& echo "\nDownloading spark-${SPARK_VERSION}-bin-hadoop${HADOOP_VERSION}.tgz ..." \`

as the spark download can take a while. Does that make sense?

indy-3rdman commented on docker/images/interactive/dotnet-spark-base/Dockerfile:
```diff
@@ -0,0 +1,25 @@
+RUN mkdir -p /dotnet/HelloSpark \
+    && mkdir -p /dotnet/Debug/netcoreapp${DOTNET_CORE_VERSION} \
+    && wget -q --show-progress --progress=bar:force:noscroll https://github.com/dotnet/spark/releases/download/v${DOTNET_SPARK_VERSION}/Microsoft.Spark.Worker.netcoreapp${DOTNET_CORE_VERSION}.linux-x64-${DOTNET_SPARK_VERSION}.tar.gz \
+    && tar -xvzf Microsoft.Spark.Worker.netcoreapp${DOTNET_CORE_VERSION}.linux-x64-${DOTNET_SPARK_VERSION}.tar.gz \
```
Should be changed in the latest commit.

indy-3rdman commented on docker/images/interactive/dotnet-spark-base/Dockerfile:
```diff
@@ -0,0 +1,25 @@
+    && wget -q --show-progress --progress=bar:force:noscroll https://github.com/dotnet/spark/releases/download/v${DOTNET_SPARK_VERSION}/Microsoft.Spark.Worker.netcoreapp${DOTNET_CORE_VERSION}.linux-x64-${DOTNET_SPARK_VERSION}.tar.gz \
+    && tar -xvzf Microsoft.Spark.Worker.netcoreapp${DOTNET_CORE_VERSION}.linux-x64-${DOTNET_SPARK_VERSION}.tar.gz \
+    && mv Microsoft.Spark.Worker-${DOTNET_SPARK_VERSION} /dotnet/ \
+    && cp /dotnet/Microsoft.Spark.Worker-${DOTNET_SPARK_VERSION}/Microsoft.Spark.Worker /dotnet/Microsoft.Spark.Worker-${DOTNET_SPARK_VERSION}/Microsoft.Spark.Worker.exe \
```
Sorry, this is just a leftover from another file. Removed.

indy-3rdman commented on docker/images/interactive/dotnet-spark-base/Dockerfile:
```diff
@@ -0,0 +1,25 @@
+    && chmod 755 /dotnet/Microsoft.Spark.Worker-${DOTNET_SPARK_VERSION}/Microsoft.Spark.Worker \
+    && rm Microsoft.Spark.Worker.netcoreapp${DOTNET_CORE_VERSION}.linux-x64-${DOTNET_SPARK_VERSION}.tar.gz
+
+COPY HelloSpark /dotnet/HelloSpark
```
Moved from dotnet-spark-base to dotnet-spark Dockerfile. Additionally this is now removed, after the jar file has been copied.

indy-3rdman commented on docker/images/interactive/dotnet-spark/Dockerfile:
```diff
@@ -1,25 +1,32 @@
+ARG DOTNET_SPARK_JAR="microsoft-spark-3-0_2.12-$DOTNET_SPARK_VERSION"
 
-USER root
+ENV DAEMON_RUN=true
```
Updated the Dockerfiles.

indy-3rdman commented on docker/images/interactive/dotnet-spark/Dockerfile:
```diff
@@ -1,25 +1,32 @@
-ARG DOTNET_SPARK_VERSION=0.12.1
-ENV DOTNET_SPARK_VERSION=$DOTNET_SPARK_VERSION
-ENV DOTNET_WORKER_DIR=/dotnet/Microsoft.Spark.Worker-${DOTNET_SPARK_VERSION}
+ARG SPARK_VERSION=3.0.1
```
This is related to my earlier point about the purpose of the Dockerfile(s). The intention was to have a build-able Dockerfile even if the build script is not used. I agree with your observation about maintenance. Maybe @rapoth has a view on that.

KimKiHyuk commented:
@indy-3rdman 
I've encountered some problem using interactive docker (jupyter)

To run my spark program, I have to pass a parameter like
```spark-submit --packages org.apache.spark:spark-avro_2.11:2.4.6 ...```

but i don't get it how to do it in jupyter.

what should i do to solve it?

```
[2020-11-26T00:28:42.3520910Z] [c2435164f8aa] [Error] [JvmBridge] org.apache.spark.sql.AnalysisException: Failed to find data source: avro. Avro is built-in but external data source module since Spark 2.4. Please deploy the application as per the deployment section of "Apache Avro Data Source Guide".;
	at org.apache.spark.sql.execution.datasources.DataSource$.lookupDataSource(DataSource.scala:666)
	at org.apache.spark.sql.execution.datasources.DataSource$.lookupDataSourceV2(DataSource.scala:733)
	at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:248)
	at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:232)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.api.dotnet.DotnetBackendHandler.handleMethodCall(DotnetBackendHandler.scala:159)
	at org.apache.spark.api.dotnet.DotnetBackendHandler.$anonfun$handleBackendRequest$2(DotnetBackendHandler.scala:99)
	at org.apache.spark.api.dotnet.ThreadPool$$anon$1.run(ThreadPool.scala:34)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```
